src/librustc_resolve/macros.rs

```diff
@@ -326,17 +325,34 @@ impl<'a> base::Resolver for Resolver<'a> {
         if let Def::Macro(_, MacroKind::ProcMacroStub) = def {
             self.report_proc_macro_stub(invoc.span());
             return Err(Determinacy::Determined);
-        } else if let Def::NonMacroAttr = def {
-            if let InvocationKind::Attr { .. } = invoc.kind {
-                if !self.session.features_untracked().tool_attributes {
-                    feature_err(&self.session.parse_sess, "tool_attributes",
-                                invoc.span(), GateIssue::Language,
-                                "tool attributes are unstable").emit();
+        } else if let Def::NonMacroAttr(attr_kind) = def {
+            let is_attr_invoc =
+                if let InvocationKind::Attr { .. } = invoc.kind { true } else { false };
+            let path = invoc.path().expect("no path for non-macro attr");
+            match attr_kind {
+                NonMacroAttrKind::Tool | NonMacroAttrKind::DeriveHelper |
+                NonMacroAttrKind::Custom if is_attr_invoc => {
```

Comment:
I'm not personally 100% familiar with resolve, so I'm a little lost here as to why `if is_attr_invoc` is needed here. Could a comment be added about why it's here and why we'd have a non-attr invocation get resolved to a `NonMacroAttr` def still?

Reply:
Anything in macro namespace can resolve to `Def::NonMacroAttr`!
```rust
let x = inline!(); // ERROR expected a macro, found built-in attribute
```